// @videojs/react/dist/dev/player/context.js
'use client';

import { useComposedRefs } from "../utils/use-composed-refs.js";
import { useStore } from "@videojs/store/react";
import { createContext, forwardRef, useContext, useEffect, useRef } from "react";
import { jsx } from "react/jsx-runtime";

//#region src/player/context.tsx
const PlayerContext = createContext(null);
const EMPTY_UNSUBSCRIBE = () => {};
const EMPTY_STORE = {
	state: {},
	subscribe: () => EMPTY_UNSUBSCRIBE
};
function PlayerContextProvider({ value, children }) {
	return /* @__PURE__ */ jsx(PlayerContext.Provider, {
		value,
		children
	});
}
/** Access the full player context value. Throws if used outside a Player Provider. */
function usePlayerContext() {
	const ctx = useContext(PlayerContext);
	if (!ctx) throw new Error("usePlayerContext must be used within a Player Provider");
	return ctx;
}
function usePlayer(selector) {
	const { store } = usePlayerContext();
	return useStore(store, selector);
}
function useOptionalPlayer(selector) {
	const ctx = useContext(PlayerContext);
	const value = useStore(ctx?.store ?? EMPTY_STORE, ctx ? selector : void 0);
	return ctx ? value : void 0;
}
/** Access the media element from within a Player Provider. */
function useMedia() {
	const { media } = usePlayerContext();
	return media;
}
/** Access the media registration setter for connecting a media element to the player. */
function useMediaRegistration() {
	return useContext(PlayerContext)?.setMedia;
}
const Container = forwardRef(function Container({ children, ...props }, ref) {
	const { store, media } = usePlayerContext();
	const internalRef = useRef(null);
	const composedRef = useComposedRefs(ref, internalRef);
	useEffect(() => {
		if (!media) return;
		return store.attach({
			media,
			container: internalRef.current
		});
	}, [media, store]);
	return /* @__PURE__ */ jsx("div", {
		ref: composedRef,
		...props,
		children
	});
});

//#endregion
export { Container, PlayerContextProvider, useMedia, useMediaRegistration, useOptionalPlayer, usePlayer, usePlayerContext };
