// @videojs/react/dist/dev/ui/time-slider/time-slider-root.js
'use client';

import { usePlayer } from "../../player/context.js";
import { renderElement } from "../../utils/use-render.js";
import { useLatestRef } from "../../utils/use-latest-ref.js";
import { useSlider } from "../hooks/use-slider.js";
import { SliderProvider } from "../slider/context.js";
import { getTimeSliderCSSVars, logMissingFeature, selectBuffer, selectTime } from "@videojs/core/dom";
import { forwardRef, useState } from "react";
import { jsx } from "react/jsx-runtime";
import { TimeSliderCore, TimeSliderDataAttrs } from "@videojs/core";
import { formatTime } from "@videojs/utils/time";

//#region src/ui/time-slider/time-slider-root.tsx
const noopSeek = () => Promise.resolve(0);
const TimeSliderRoot = forwardRef(function TimeSliderRoot(componentProps, forwardedRef) {
	const { render, className, style, label, commitThrottle = TimeSliderCore.defaultProps.commitThrottle, step = TimeSliderCore.defaultProps.step, largeStep = TimeSliderCore.defaultProps.largeStep, orientation, disabled, thumbAlignment, onDragStart, onDragEnd, ...elementProps } = componentProps;
	const time = usePlayer(selectTime);
	const buffer = usePlayer(selectBuffer);
	const [core] = useState(() => new TimeSliderCore());
	core.setProps({
		label,
		step,
		largeStep,
		orientation,
		disabled,
		thumbAlignment
	});
	const mediaRef = useLatestRef(time && buffer ? {
		...time,
		...buffer
	} : null);
	const duration = time?.duration ?? 0;
	const { state, cssVars, rootRef, thumbRef, rootProps, thumbProps } = useSlider({
		computeState: (input) => {
			core.setInput(input);
			if (!time || !buffer) core.setMedia({
				currentTime: 0,
				duration: 0,
				seeking: false,
				seek: noopSeek,
				buffered: [],
				seekable: []
			});
			else core.setMedia({
				...time,
				...buffer
			});
			return core.getState();
		},
		getPercent: () => duration > 0 ? (time?.currentTime ?? 0) / duration * 100 : 0,
		getStepPercent: () => core.getStepPercent(),
		getLargeStepPercent: () => core.getLargeStepPercent(),
		orientation,
		disabled,
		commitThrottle,
		adjustPercent: (rawPercent, thumbSize, trackSize) => core.adjustPercentForAlignment(rawPercent, thumbSize, trackSize),
		getCSSVars: getTimeSliderCSSVars,
		onValueCommit: (percent) => {
			const media = mediaRef.current;
			if (media) media.seek(core.valueFromPercent(percent));
		},
		onDragStart,
		onDragEnd
	});
	if (!time) {
		logMissingFeature("TimeSlider", "time");
		return null;
	}
	return /* @__PURE__ */ jsx(SliderProvider, {
		value: {
			state,
			pointerValue: core.valueFromPercent(state.pointerPercent),
			thumbRef,
			thumbProps,
			stateAttrMap: TimeSliderDataAttrs,
			getAttrs: (sliderState) => core.getAttrs(sliderState),
			formatValue: (value) => formatTime(value, duration)
		},
		children: renderElement("div", {
			render,
			className,
			style
		}, {
			state,
			stateAttrMap: TimeSliderDataAttrs,
			ref: [forwardedRef, rootRef],
			props: [
				{ style: cssVars },
				rootProps,
				elementProps
			]
		})
	});
});

//#endregion
export { TimeSliderRoot };
