// @videojs/core/dist/default/core/ui/pip-button/pip-button-core.js
import { defaults } from "@videojs/utils/object";
import { isFunction } from "@videojs/utils/predicate";

//#region src/core/ui/pip-button/pip-button-core.ts
var PiPButtonCore = class PiPButtonCore {
	static defaultProps = {
		label: "",
		disabled: false
	};
	#props = { ...PiPButtonCore.defaultProps };
	#media = null;
	constructor(props) {
		if (props) this.setProps(props);
	}
	setProps(props) {
		this.#props = defaults(props, PiPButtonCore.defaultProps);
	}
	getLabel(state) {
		const { label } = this.#props;
		if (isFunction(label)) {
			const customLabel = label(state);
			if (customLabel) return customLabel;
		} else if (label) return label;
		return state.pip ? "Exit picture-in-picture" : "Enter picture-in-picture";
	}
	getAttrs(state) {
		return {
			"aria-label": this.getLabel(state),
			"aria-disabled": this.#props.disabled ? "true" : void 0
		};
	}
	setMedia(media) {
		this.#media = media;
	}
	getState() {
		const media = this.#media;
		return {
			pip: media.pip,
			availability: media.pipAvailability
		};
	}
	async toggle(media) {
		if (this.#props.disabled) return;
		if (media.pipAvailability !== "available") return;
		try {
			if (media.pip) await media.exitPiP();
			else await media.requestPiP();
		} catch {}
	}
};

//#endregion
export { PiPButtonCore };
