// @videojs/react/dist/default/ui/popover/popover-popup.js
'use client';

import { useComposedRefs } from "../../utils/use-composed-refs.js";
import { renderElement } from "../../utils/use-render.js";
import { usePopoverContext } from "./context.js";
import { getAnchorPositionStyle, resolveOffsets } from "@videojs/core/dom";
import { forwardRef, useCallback, useLayoutEffect, useMemo, useRef, useState } from "react";
import { supportsAnchorPositioning } from "@videojs/utils/dom";

//#region src/ui/popover/popover-popup.tsx
const POPOVER_RESET = {
	position: "fixed",
	inset: "auto",
	margin: 0
};
/** Container for the popover content. Positioned relative to the trigger using CSS anchor positioning with a JavaScript fallback. */
const PopoverPopup = forwardRef(function PopoverPopup({ render, className, style, ...elementProps }, forwardedRef) {
	const { core, popover, state, stateAttrMap, anchorName, popupId } = usePopoverContext();
	const internalRef = useRef(null);
	const composedRef = useComposedRefs(forwardedRef, useCallback((el) => {
		popover.setPopupElement(el);
	}, [popover]), internalRef);
	const posOpts = useMemo(() => ({
		side: state.side,
		align: state.align
	}), [state.side, state.align]);
	const anchorStyle = useMemo(() => {
		if (!supportsAnchorPositioning()) return null;
		return getAnchorPositionStyle(anchorName, posOpts);
	}, [anchorName, posOpts]);
	const [manualStyle, setManualStyle] = useState(null);
	useLayoutEffect(() => {
		if (supportsAnchorPositioning()) return;
		if (!state.open) {
			setManualStyle(null);
			return;
		}
		function measure() {
			const triggerEl = popover.triggerElement;
			const popupEl = internalRef.current;
			if (!triggerEl || !popupEl) return;
			setManualStyle(getAnchorPositionStyle(anchorName, posOpts, triggerEl.getBoundingClientRect(), popupEl.getBoundingClientRect(), document.documentElement.getBoundingClientRect(), resolveOffsets(popupEl)));
		}
		measure();
		let rafId = 0;
		function reposition() {
			cancelAnimationFrame(rafId);
			rafId = requestAnimationFrame(measure);
		}
		window.addEventListener("scroll", reposition, {
			capture: true,
			passive: true
		});
		window.addEventListener("resize", reposition);
		return () => {
			cancelAnimationFrame(rafId);
			window.removeEventListener("scroll", reposition, true);
			window.removeEventListener("resize", reposition);
		};
	}, [
		state.open,
		anchorName,
		posOpts,
		popover
	]);
	const positioningStyle = anchorStyle ?? manualStyle ?? POPOVER_RESET;
	if (!state.open) return null;
	const { onFocusOut, ...restPopupProps } = popover.popupProps;
	return renderElement("div", {
		render,
		className,
		style
	}, {
		state,
		stateAttrMap,
		ref: composedRef,
		props: [
			{
				id: popupId,
				style: positioningStyle,
				...core.getPopupAttrs(state)
			},
			{
				...restPopupProps,
				onBlur: onFocusOut
			},
			elementProps
		]
	});
});

//#endregion
export { PopoverPopup };
