// @videojs/react/dist/default/ui/popover/popover-root.js
'use client';

import { useLatestRef } from "../../utils/use-latest-ref.js";
import { PopoverContextProvider } from "./context.js";
import { useSafeId } from "../../utils/use-safe-id.js";
import { createPopover, createTransition } from "@videojs/core/dom";
import { useSnapshot } from "@videojs/store/react";
import { useEffect, useState } from "react";
import { jsx } from "react/jsx-runtime";
import { PopoverCore, PopoverDataAttrs } from "@videojs/core";

//#region src/ui/popover/popover-root.tsx
function PopoverRoot({ open: controlledOpen, defaultOpen = PopoverCore.defaultProps.defaultOpen, onOpenChange: onOpenChangeProp, onOpenChangeComplete: onOpenChangeCompleteProp, openOnHover = PopoverCore.defaultProps.openOnHover, delay = PopoverCore.defaultProps.delay, closeDelay = PopoverCore.defaultProps.closeDelay, children, ...coreProps }) {
	const [core] = useState(() => new PopoverCore(coreProps));
	core.setProps(coreProps);
	const isControlled = controlledOpen !== void 0;
	const onOpenChangeRef = useLatestRef(onOpenChangeProp);
	const onOpenChangeCompleteRef = useLatestRef(onOpenChangeCompleteProp);
	const closeOnEscapeRef = useLatestRef(coreProps.closeOnEscape);
	const closeOnOutsideClickRef = useLatestRef(coreProps.closeOnOutsideClick);
	const openOnHoverRef = useLatestRef(openOnHover);
	const delayRef = useLatestRef(delay);
	const closeDelayRef = useLatestRef(closeDelay);
	const [popover] = useState(() => {
		const instance = createPopover({
			transition: createTransition(),
			onOpenChange: (nextOpen, details) => {
				onOpenChangeRef.current?.(nextOpen, details);
			},
			onOpenChangeComplete: (nextOpen) => {
				onOpenChangeCompleteRef.current?.(nextOpen);
			},
			closeOnEscape: () => closeOnEscapeRef.current ?? PopoverCore.defaultProps.closeOnEscape,
			closeOnOutsideClick: () => closeOnOutsideClickRef.current ?? PopoverCore.defaultProps.closeOnOutsideClick,
			openOnHover: () => openOnHoverRef.current,
			delay: () => delayRef.current,
			closeDelay: () => closeDelayRef.current
		});
		if (!isControlled && defaultOpen) instance.open("click");
		return instance;
	});
	const anchorName = useSafeId();
	const popupId = useSafeId("popup-");
	useEffect(() => {
		if (controlledOpen === void 0) return;
		const { active: inputOpen } = popover.input.current;
		if (controlledOpen === inputOpen) return;
		if (controlledOpen) popover.open("click");
		else popover.close("click");
	}, [controlledOpen, popover]);
	useEffect(() => () => popover.destroy(), [popover]);
	const input = useSnapshot(popover.input);
	core.setInput(input);
	return /* @__PURE__ */ jsx(PopoverContextProvider, {
		value: {
			core,
			popover,
			state: core.getState(),
			stateAttrMap: PopoverDataAttrs,
			anchorName,
			popupId
		},
		children
	});
}

//#endregion
export { PopoverRoot };
