// @videojs/core/dist/default/dom/ui/slider.js
import { getPercentFromPointerEvent } from "../utils/pointer.js";
import { createState } from "@videojs/store";
import { isNull } from "@videojs/utils/predicate";
import { clamp, roundToStep } from "@videojs/utils/number";
import { listen } from "@videojs/utils/dom";
import { throttle } from "@videojs/utils/function";

//#region src/dom/ui/slider.ts
/** Intentional drag threshold — number of pointermove events before drag starts. */
const DRAG_THRESHOLD = 2;
function createSlider(options) {
	const input = createState({
		pointerPercent: 0,
		dragPercent: 0,
		dragging: false,
		pointing: false,
		focused: false
	});
	const abort = new AbortController();
	const commitThrottleMs = options.commitThrottle ?? 0;
	let isDragging = false, moveCount = 0, cachedRTL = false, cachedRect = null, documentCleanup = null, capturedPointerId = null;
	const throttledCommit = commitThrottleMs > 0 ? throttle((percent) => options.onValueCommit?.(percent), commitThrottleMs) : null;
	function releaseCapture() {
		if (isNull(capturedPointerId)) return;
		const id = capturedPointerId;
		capturedPointerId = null;
		try {
			options.getElement().releasePointerCapture(id);
		} catch {}
	}
	function endDrag() {
		if (!isDragging) input.patch({
			pointing: false,
			pointerPercent: 0
		});
		else {
			isDragging = false;
			input.patch({
				dragging: false,
				pointing: false,
				pointerPercent: 0
			});
			options.onDragEnd?.();
		}
		cleanup();
	}
	function cleanup() {
		throttledCommit?.cancel();
		releaseCapture();
		documentCleanup?.();
		documentCleanup = null;
		cachedRect = null;
	}
	function onDocumentPointerMove(event) {
		if (event.pointerType !== "touch" && event.buttons === 0) {
			endDrag();
			return;
		}
		moveCount++;
		const percent = getPercentFromPointerEvent(event, cachedRect, options.getOrientation(), cachedRTL);
		if (!isDragging && moveCount >= DRAG_THRESHOLD) {
			isDragging = true;
			input.patch({
				dragging: true,
				dragPercent: percent,
				pointerPercent: percent
			});
			options.onDragStart?.();
			options.onValueChange?.(percent);
			throttledCommit?.(percent);
		} else if (isDragging) {
			input.patch({
				dragPercent: percent,
				pointerPercent: percent
			});
			options.onValueChange?.(percent);
			throttledCommit?.(percent);
		} else input.patch({ pointerPercent: percent });
	}
	function onDocumentPointerUp(event) {
		const percent = getPercentFromPointerEvent(event, cachedRect, options.getOrientation(), cachedRTL);
		throttledCommit?.cancel();
		options.onValueCommit?.(percent);
		endDrag();
	}
	function addDocumentListeners() {
		const abort = new AbortController();
		const signal = abort.signal;
		listen(document, "pointermove", onDocumentPointerMove, {
			passive: true,
			signal
		});
		listen(document, "pointerup", onDocumentPointerUp, { signal });
		listen(document, "pointercancel", endDrag, { signal });
		listen(document, "touchmove", (e) => e.preventDefault(), {
			passive: false,
			signal
		});
		documentCleanup = () => abort.abort();
	}
	const rootProps = {
		onPointerDown(event) {
			if (options.isDisabled()) return;
			const el = options.getElement();
			cachedRect = el.getBoundingClientRect();
			cachedRTL = options.isRTL();
			moveCount = 0;
			releaseCapture();
			capturedPointerId = event.pointerId;
			el.setPointerCapture(event.pointerId);
			const percent = getPercentFromPointerEvent(event, cachedRect, options.getOrientation(), cachedRTL);
			input.patch({
				pointing: true,
				pointerPercent: percent,
				dragPercent: percent
			});
			options.onValueChange?.(percent);
			options.getThumbElement?.()?.focus();
			documentCleanup?.();
			addDocumentListeners();
		},
		onPointerMove(event) {
			if (options.isDisabled() || isDragging) return;
			const percent = getPercentFromPointerEvent(event, options.getElement().getBoundingClientRect(), options.getOrientation(), options.isRTL());
			input.patch({
				pointing: true,
				pointerPercent: percent
			});
		},
		onPointerLeave() {
			if (isDragging) return;
			input.patch({
				pointing: false,
				pointerPercent: 0
			});
		}
	};
	const thumbProps = {
		onKeyDown(event) {
			if (options.isDisabled()) {
				if (event.key !== "Tab") event.preventDefault();
				return;
			}
			const stepPercent = options.getStepPercent();
			const largeStepPercent = options.getLargeStepPercent();
			const rounded = roundToStep(options.getPercent(), stepPercent, 0);
			const horizontalSign = options.isRTL() ? -1 : 1;
			const step = event.shiftKey ? largeStepPercent : stepPercent;
			let newPercent = null;
			switch (event.key) {
				case "ArrowRight":
					newPercent = rounded + step * horizontalSign;
					break;
				case "ArrowLeft":
					newPercent = rounded - step * horizontalSign;
					break;
				case "ArrowUp":
					newPercent = rounded + step;
					break;
				case "ArrowDown":
					newPercent = rounded - step;
					break;
				case "PageUp":
					newPercent = rounded + largeStepPercent;
					break;
				case "PageDown":
					newPercent = rounded - largeStepPercent;
					break;
				case "Home":
					newPercent = 0;
					break;
				case "End":
					newPercent = 100;
					break;
				default:
					if (!event.metaKey && !event.ctrlKey && !event.altKey && event.key >= "0" && event.key <= "9") newPercent = Number(event.key) * 10;
					break;
			}
			if (newPercent !== null) {
				event.preventDefault();
				newPercent = clamp(newPercent, 0, 100);
				input.patch({
					pointerPercent: newPercent,
					dragPercent: newPercent
				});
				options.onValueChange?.(newPercent);
				options.onValueCommit?.(newPercent);
			}
		},
		onFocus() {
			input.patch({ focused: true });
		},
		onBlur() {
			input.patch({ focused: false });
		}
	};
	listen(abort.signal, "abort", cleanup, { once: true });
	return {
		input,
		rootProps,
		thumbProps,
		destroy() {
			abort.abort();
		}
	};
}

//#endregion
export { createSlider };
