// @videojs/core/dist/default/dom/ui/popover/popover-positioning.js
import { PopoverCSSVars } from "../../../core/ui/popover/popover-css-vars.js";
import { supportsAnchorPositioning } from "@videojs/utils/dom";

//#region src/dom/ui/popover/popover-positioning.ts
const OPPOSITE_SIDE = {
	top: "bottom",
	bottom: "top",
	left: "right",
	right: "left"
};
/**
* Get positioning styles for the popup element.
*
* When the browser supports CSS Anchor Positioning, returns native CSS properties
* that reference `var(--media-popover-side-offset, 0px)` and
* `var(--media-popover-align-offset, 0px)` — no JS offset values needed.
*
* When rects are provided and anchor positioning is unsupported, falls back to
* manual JS-computed positioning. The caller must resolve offset CSS vars via
* `getComputedStyle` and pass them as `offsets`.
*
* Returns camelCase keys for standard CSS properties and `--*` keys for
* custom properties — compatible with both React's `style` prop and
* `applyStyles()` from `@videojs/utils/dom`.
*/
function getAnchorPositionStyle(anchorName, opts, triggerRect, popupRect, boundaryRect, offsets) {
	if (supportsAnchorPositioning()) return getAnchorPositionCSS(anchorName, opts);
	if (triggerRect && popupRect) return {
		...getManualPositionStyle(triggerRect, popupRect, opts, offsets ?? {
			sideOffset: 0,
			alignOffset: 0
		}),
		...boundaryRect ? getPopoverCSSVars(triggerRect, boundaryRect, opts.side) : {},
		position: "fixed",
		inset: "auto",
		margin: "0"
	};
	return {};
}
/** Generate style to set on the trigger for CSS Anchor Positioning. */
function getAnchorNameStyle(anchorName) {
	if (!supportsAnchorPositioning()) return {};
	return { anchorName: `--${anchorName}` };
}
const SIDE_OFFSET_VAR = `var(${PopoverCSSVars.sideOffset}, 0px)`;
const ALIGN_OFFSET_VAR = `var(${PopoverCSSVars.alignOffset}, 0px)`;
function getAnchorPositionCSS(anchorName, opts) {
	const { side, align } = opts;
	const style = {
		positionAnchor: `--${anchorName}`,
		position: "fixed",
		inset: "auto",
		margin: "0",
		justifySelf: "normal",
		alignSelf: "normal",
		marginInlineStart: "0",
		marginBlockStart: "0"
	};
	const insetProp = OPPOSITE_SIDE[side];
	if (side === "top" || side === "bottom") {
		style[insetProp] = `calc(anchor(${side}) + ${SIDE_OFFSET_VAR})`;
		if (align === "start") style.left = `calc(anchor(left) + ${ALIGN_OFFSET_VAR})`;
		else if (align === "end") style.right = `calc(anchor(right) + ${ALIGN_OFFSET_VAR})`;
		else {
			style.justifySelf = "anchor-center";
			style.marginInlineStart = ALIGN_OFFSET_VAR;
		}
	} else {
		style[insetProp] = `calc(anchor(${side}) + ${SIDE_OFFSET_VAR})`;
		if (align === "start") style.top = `calc(anchor(top) + ${ALIGN_OFFSET_VAR})`;
		else if (align === "end") style.bottom = `calc(anchor(bottom) + ${ALIGN_OFFSET_VAR})`;
		else {
			style.alignSelf = "anchor-center";
			style.marginBlockStart = ALIGN_OFFSET_VAR;
		}
	}
	return style;
}
/**
* Compute CSS variables for the popup element.
*
* These enable CSS-based sizing constraints relative to the anchor/boundary.
*/
function getPopoverCSSVars(triggerRect, boundaryRect, side) {
	const vars = {};
	vars[PopoverCSSVars.anchorWidth] = `${triggerRect.width}px`;
	vars[PopoverCSSVars.anchorHeight] = `${triggerRect.height}px`;
	if (side === "top" || side === "bottom") {
		vars[PopoverCSSVars.availableHeight] = side === "top" ? `${triggerRect.top - boundaryRect.top}px` : `${boundaryRect.bottom - triggerRect.bottom}px`;
		vars[PopoverCSSVars.availableWidth] = `${boundaryRect.width}px`;
	} else {
		vars[PopoverCSSVars.availableWidth] = side === "left" ? `${triggerRect.left - boundaryRect.left}px` : `${boundaryRect.right - triggerRect.right}px`;
		vars[PopoverCSSVars.availableHeight] = `${boundaryRect.height}px`;
	}
	return vars;
}
/**
* Compute manual positioning when CSS Anchor Positioning is not supported.
*
* Returns inline `top`/`left` styles in **viewport coordinates** for use
* with `position: fixed` (the popup is in the top layer). All rects from
* `getBoundingClientRect()` are already viewport-relative.
*
* Offsets are resolved by the caller from CSS custom properties via
* `getComputedStyle()` and passed as `offsets`.
*/
function getManualPositionStyle(triggerRect, popupRect, opts, offsets = {
	sideOffset: 0,
	alignOffset: 0
}) {
	const { side, align } = opts;
	const { sideOffset, alignOffset } = offsets;
	let top = 0;
	let left = 0;
	if (side === "top") top = triggerRect.top - popupRect.height - sideOffset;
	else if (side === "bottom") top = triggerRect.bottom + sideOffset;
	else if (side === "left") left = triggerRect.left - popupRect.width - sideOffset;
	else left = triggerRect.right + sideOffset;
	if (side === "top" || side === "bottom") if (align === "start") left = triggerRect.left + alignOffset;
	else if (align === "end") left = triggerRect.right - popupRect.width + alignOffset;
	else left = triggerRect.left + (triggerRect.width - popupRect.width) / 2 + alignOffset;
	else if (align === "start") top = triggerRect.top + alignOffset;
	else if (align === "end") top = triggerRect.bottom - popupRect.height + alignOffset;
	else top = triggerRect.top + (triggerRect.height - popupRect.height) / 2 + alignOffset;
	return {
		top: `${top}px`,
		left: `${left}px`
	};
}
/**
* Read `--media-popover-side-offset` and `--media-popover-align-offset`
* from the popup element's computed style, returning numeric pixel values.
*/
function resolveOffsets(el) {
	const computed = getComputedStyle(el);
	return {
		sideOffset: Number.parseFloat(computed.getPropertyValue(PopoverCSSVars.sideOffset)) || 0,
		alignOffset: Number.parseFloat(computed.getPropertyValue(PopoverCSSVars.alignOffset)) || 0
	};
}

//#endregion
export { getAnchorNameStyle, getAnchorPositionStyle, getManualPositionStyle, getPopoverCSSVars, resolveOffsets };
