// @videojs/react/dist/default/presets/video/minimal-skin.js
import { Container } from "../../player/context.js";
import { BufferingIndicator } from "../../ui/buffering-indicator/buffering-indicator.js";
import { ControlsRoot } from "../../ui/controls/controls-root.js";
import { FullscreenButton } from "../../ui/fullscreen-button/fullscreen-button.js";
import { MuteButton } from "../../ui/mute-button/mute-button.js";
import { PiPButton } from "../../ui/pip-button/pip-button.js";
import { PlayButton } from "../../ui/play-button/play-button.js";
import { PlaybackRateButton } from "../../ui/playback-rate-button/playback-rate-button.js";
import { PopoverPopup } from "../../ui/popover/popover-popup.js";
import { PopoverRoot } from "../../ui/popover/popover-root.js";
import { PopoverTrigger } from "../../ui/popover/popover-trigger.js";
import { SeekButton } from "../../ui/seek-button/seek-button.js";
import { SliderBuffer } from "../../ui/slider/slider-buffer.js";
import { SliderFill } from "../../ui/slider/slider-fill.js";
import { SliderThumb } from "../../ui/slider/slider-thumb.js";
import { SliderTrack } from "../../ui/slider/slider-track.js";
import { Group } from "../../ui/time/time-group.js";
import { Separator } from "../../ui/time/time-separator.js";
import { Value } from "../../ui/time/time-value.js";
import { TimeSliderRoot } from "../../ui/time-slider/time-slider-root.js";
import { VolumeSliderRoot } from "../../ui/volume-slider/volume-slider-root.js";
import CaptionsOffIcon from "../../icons/dist/react/minimal/captions-off.js";
import CaptionsOnIcon from "../../icons/dist/react/minimal/captions-on.js";
import FullscreenEnterIcon from "../../icons/dist/react/minimal/fullscreen-enter.js";
import FullscreenExitIcon from "../../icons/dist/react/minimal/fullscreen-exit.js";
import PauseIcon from "../../icons/dist/react/minimal/pause.js";
import PipIcon from "../../icons/dist/react/minimal/pip.js";
import PlayIcon from "../../icons/dist/react/minimal/play.js";
import RestartIcon from "../../icons/dist/react/minimal/restart.js";
import SeekIcon from "../../icons/dist/react/minimal/seek.js";
import SpinnerIcon from "../../icons/dist/react/minimal/spinner.js";
import VolumeHighIcon from "../../icons/dist/react/minimal/volume-high.js";
import VolumeLowIcon from "../../icons/dist/react/minimal/volume-low.js";
import VolumeOffIcon from "../../icons/dist/react/minimal/volume-off.js";
import { CaptionsButton } from "../../ui/captions-button/captions-button.js";
import { ErrorDialog } from "../../ui/error-dialog/error-dialog.js";
import { forwardRef } from "react";
import { jsx, jsxs } from "react/jsx-runtime";
import { cn } from "@videojs/utils/style";

//#region src/presets/video/minimal-skin.tsx
const SEEK_TIME = 10;
const Button = forwardRef(function Button({ className, ...props }, ref) {
	return /* @__PURE__ */ jsx("button", {
		ref,
		type: "button",
		className: cn("media-button", className),
		...props
	});
});
function MinimalVideoSkin(props) {
	const { children, className, ...rest } = props;
	return /* @__PURE__ */ jsxs(Container, {
		className: cn("media-minimal-skin", className),
		...rest,
		children: [
			/* @__PURE__ */ jsx(BufferingIndicator, { render: (props) => /* @__PURE__ */ jsx("div", {
				...props,
				className: "media-buffering-indicator",
				children: /* @__PURE__ */ jsx(SpinnerIcon, { className: "media-icon" })
			}) }),
			/* @__PURE__ */ jsx(ErrorDialog, {
				"aria-labelledby": "media-error-title",
				"aria-describedby": "media-error-description",
				render: (props, { onDismiss }) => /* @__PURE__ */ jsx("div", {
					...props,
					className: "media-error",
					children: /* @__PURE__ */ jsxs("div", {
						className: "media-error__dialog",
						children: [/* @__PURE__ */ jsxs("div", {
							className: "media-error__content",
							children: [/* @__PURE__ */ jsx("p", {
								id: "media-error-title",
								className: "media-error__title",
								children: "Something went wrong."
							}), /* @__PURE__ */ jsx("p", {
								id: "media-error-description",
								className: "media-error__description",
								children: "An error occurred while trying to play the video. Please try again."
							})]
						}), /* @__PURE__ */ jsx("div", {
							className: "media-error__actions",
							children: /* @__PURE__ */ jsx(Button, {
								onClick: onDismiss,
								children: "OK"
							})
						})]
					})
				})
			}),
			/* @__PURE__ */ jsxs(ControlsRoot, {
				className: "media-controls",
				children: [
					/* @__PURE__ */ jsxs("span", {
						className: "media-button-group",
						children: [
							/* @__PURE__ */ jsx(PlayButton, { render: (props) => /* @__PURE__ */ jsxs(Button, {
								...props,
								className: "media-button--icon media-button--play",
								children: [
									/* @__PURE__ */ jsx(RestartIcon, { className: "media-icon media-icon--restart" }),
									/* @__PURE__ */ jsx(PlayIcon, { className: "media-icon media-icon--play" }),
									/* @__PURE__ */ jsx(PauseIcon, { className: "media-icon media-icon--pause" })
								]
							}) }),
							/* @__PURE__ */ jsx(SeekButton, {
								seconds: -SEEK_TIME,
								render: (props) => /* @__PURE__ */ jsx(Button, {
									...props,
									className: "media-button--icon media-button--seek",
									children: /* @__PURE__ */ jsxs("span", {
										className: "media-icon__container",
										children: [/* @__PURE__ */ jsx(SeekIcon, { className: "media-icon media-icon--seek media-icon--flipped" }), /* @__PURE__ */ jsx("span", {
											className: "media-icon__label",
											children: SEEK_TIME
										})]
									})
								})
							}),
							/* @__PURE__ */ jsx(SeekButton, {
								seconds: SEEK_TIME,
								render: (props) => /* @__PURE__ */ jsx(Button, {
									...props,
									className: "media-button--icon media-button--seek",
									children: /* @__PURE__ */ jsxs("span", {
										className: "media-icon__container",
										children: [/* @__PURE__ */ jsx(SeekIcon, { className: "media-icon media-icon--seek" }), /* @__PURE__ */ jsx("span", {
											className: "media-icon__label",
											children: SEEK_TIME
										})]
									})
								})
							})
						]
					}),
					/* @__PURE__ */ jsxs("span", {
						className: "media-time-controls",
						children: [/* @__PURE__ */ jsxs(Group, {
							className: "media-time",
							children: [
								/* @__PURE__ */ jsx(Value, {
									type: "current",
									className: "media-time__value media-time__value--current"
								}),
								/* @__PURE__ */ jsx(Separator, { className: "media-time__separator" }),
								/* @__PURE__ */ jsx(Value, {
									type: "duration",
									className: "media-time__value media-time__value--duration"
								})
							]
						}), /* @__PURE__ */ jsxs(TimeSliderRoot, {
							className: "media-slider",
							children: [/* @__PURE__ */ jsxs(SliderTrack, {
								className: "media-slider__track",
								children: [/* @__PURE__ */ jsx(SliderFill, { className: "media-slider__fill" }), /* @__PURE__ */ jsx(SliderBuffer, { className: "media-slider__buffer" })]
							}), /* @__PURE__ */ jsx(SliderThumb, { className: "media-slider__thumb" })]
						})]
					}),
					/* @__PURE__ */ jsxs("span", {
						className: "media-button-group",
						children: [
							/* @__PURE__ */ jsx(PlaybackRateButton, { render: (props) => /* @__PURE__ */ jsx(Button, {
								...props,
								className: "media-button--icon media-button--playback-rate"
							}) }),
							/* @__PURE__ */ jsxs(PopoverRoot, {
								openOnHover: true,
								delay: 200,
								closeDelay: 100,
								side: "top",
								children: [/* @__PURE__ */ jsx(PopoverTrigger, { render: /* @__PURE__ */ jsx(MuteButton, { render: (props) => /* @__PURE__ */ jsxs(Button, {
									...props,
									className: "media-button--icon media-button--mute",
									children: [
										/* @__PURE__ */ jsx(VolumeOffIcon, { className: "media-icon media-icon--volume-off" }),
										/* @__PURE__ */ jsx(VolumeLowIcon, { className: "media-icon media-icon--volume-low" }),
										/* @__PURE__ */ jsx(VolumeHighIcon, { className: "media-icon media-icon--volume-high" })
									]
								}) }) }), /* @__PURE__ */ jsx(PopoverPopup, {
									className: "media-surface media-popup media-popup--volume media-popup-animation",
									children: /* @__PURE__ */ jsxs(VolumeSliderRoot, {
										className: "media-slider",
										orientation: "vertical",
										thumbAlignment: "edge",
										children: [/* @__PURE__ */ jsx(SliderTrack, {
											className: "media-slider__track",
											children: /* @__PURE__ */ jsx(SliderFill, { className: "media-slider__fill" })
										}), /* @__PURE__ */ jsx(SliderThumb, { className: "media-slider__thumb media-slider__thumb--persistent" })]
									})
								})]
							}),
							/* @__PURE__ */ jsx(CaptionsButton, { render: (props) => /* @__PURE__ */ jsxs(Button, {
								...props,
								className: "media-button--icon media-button--captions",
								children: [/* @__PURE__ */ jsx(CaptionsOffIcon, { className: "media-icon media-icon--captions-off" }), /* @__PURE__ */ jsx(CaptionsOnIcon, { className: "media-icon media-icon--captions-on" })]
							}) }),
							/* @__PURE__ */ jsx(PiPButton, { render: (props) => /* @__PURE__ */ jsx(Button, {
								...props,
								className: "media-button--icon",
								children: /* @__PURE__ */ jsx(PipIcon, { className: "media-icon" })
							}) }),
							/* @__PURE__ */ jsx(FullscreenButton, { render: (props) => /* @__PURE__ */ jsxs(Button, {
								...props,
								className: "media-button--icon media-button--fullscreen",
								children: [/* @__PURE__ */ jsx(FullscreenEnterIcon, { className: "media-icon media-icon--fullscreen-enter" }), /* @__PURE__ */ jsx(FullscreenExitIcon, { className: "media-icon media-icon--fullscreen-exit" })]
							}) })
						]
					})
				]
			}),
			/* @__PURE__ */ jsx("div", { className: "media-overlay" }),
			children
		]
	});
}

//#endregion
export { MinimalVideoSkin };
