// @videojs/core/dist/default/dom.js
import { definePlayerFeature } from "./dom/feature.js";
import { bufferFeature } from "./dom/store/features/buffer.js";
import { controlsFeature } from "./dom/store/features/controls.js";
import { fullscreenFeature } from "./dom/store/features/fullscreen.js";
import { pipFeature } from "./dom/store/features/pip.js";
import { playbackFeature } from "./dom/store/features/playback.js";
import { playbackRateFeature } from "./dom/store/features/playback-rate.js";
import { sourceFeature } from "./dom/store/features/source.js";
import { textTrackFeature } from "./dom/store/features/text-track.js";
import { timeFeature } from "./dom/store/features/time.js";
import { volumeFeature } from "./dom/store/features/volume.js";
import { audioFeatures, backgroundFeatures, videoFeatures } from "./dom/store/features/presets.js";
import { feature_parts_exports } from "./dom/store/features/feature.parts.js";
import "./dom/store/features/index.js";
import { selectBuffer, selectControls, selectFullscreen, selectPiP, selectPlayback, selectPlaybackRate, selectSource, selectTextTrack, selectTime, selectVolume } from "./dom/store/selectors.js";
import { createButton } from "./dom/ui/button.js";
import { createPopover } from "./dom/ui/popover/popover.js";
import { getAnchorNameStyle, getAnchorPositionStyle, getManualPositionStyle, getPopoverCSSVars, resolveOffsets } from "./dom/ui/popover/popover-positioning.js";
import { getPercentFromPointerEvent } from "./dom/utils/pointer.js";
import { createSlider } from "./dom/ui/slider.js";
import { getSliderCSSVars, getTimeSliderCSSVars } from "./dom/ui/slider-css-vars.js";
import { createThumbnail } from "./dom/ui/thumbnail.js";
import { createTransition } from "./dom/ui/transition.js";
import { applyElementProps } from "./dom/utils/element-props.js";
import { logMissingFeature } from "./dom/utils/log.js";
import { applyStateDataAttrs, getStateDataAttrs } from "./dom/utils/state-data-attrs.js";
import "./dom/utils/index.js";

export { applyElementProps, applyStateDataAttrs, audioFeatures, backgroundFeatures, bufferFeature, controlsFeature, createButton, createPopover, createSlider, createThumbnail, createTransition, definePlayerFeature, feature_parts_exports as features, fullscreenFeature, getAnchorNameStyle, getAnchorPositionStyle, getManualPositionStyle, getPercentFromPointerEvent, getPopoverCSSVars, getSliderCSSVars, getStateDataAttrs, getTimeSliderCSSVars, logMissingFeature, pipFeature, playbackFeature, playbackRateFeature, resolveOffsets, selectBuffer, selectControls, selectFullscreen, selectPiP, selectPlayback, selectPlaybackRate, selectSource, selectTextTrack, selectTime, selectVolume, sourceFeature, textTrackFeature, timeFeature, videoFeatures, volumeFeature };
