// @videojs/core/dist/default/dom/store/features/controls.js
import { definePlayerFeature } from "../../feature.js";
import { isNull } from "@videojs/utils/predicate";
import { listen } from "@videojs/utils/dom";

//#region src/dom/store/features/controls.ts
const IDLE_DELAY = 2e3;
const TAP_THRESHOLD = 250;
const controlsFeature = definePlayerFeature({
	name: "controls",
	state: () => ({
		userActive: true,
		controlsVisible: true
	}),
	attach({ target, signal, get, set }) {
		const { media, container } = target;
		if (isNull(container)) return;
		function computeVisible(userActive) {
			return userActive || media.paused;
		}
		let idleTimer;
		function clearIdle() {
			clearTimeout(idleTimer);
			idleTimer = void 0;
		}
		function scheduleIdle() {
			clearIdle();
			idleTimer = setTimeout(setInactive, IDLE_DELAY);
		}
		function setActive() {
			if (!get().userActive) set({
				userActive: true,
				controlsVisible: true
			});
			scheduleIdle();
		}
		function setInactive() {
			clearIdle();
			set({
				userActive: false,
				controlsVisible: computeVisible(false)
			});
		}
		let pointerDownTime = 0;
		function onPointerDown() {
			pointerDownTime = Date.now();
		}
		function onPointerUp(event) {
			if (event.pointerType === "touch" && Date.now() - pointerDownTime < TAP_THRESHOLD) if (get().controlsVisible) {
				clearIdle();
				set({
					userActive: false,
					controlsVisible: computeVisible(false)
				});
			} else setActive();
			else setActive();
		}
		function onPlaybackChange() {
			const { userActive } = get();
			set({ controlsVisible: computeVisible(userActive) });
			if (!media.paused && userActive) scheduleIdle();
		}
		listen(container, "pointermove", setActive, { signal });
		listen(container, "pointerdown", onPointerDown, { signal });
		listen(container, "pointerup", onPointerUp, { signal });
		listen(container, "keyup", setActive, { signal });
		listen(container, "focusin", setActive, { signal });
		listen(container, "pointerleave", setInactive, { signal });
		listen(media, "play", onPlaybackChange, { signal });
		listen(media, "pause", onPlaybackChange, { signal });
		listen(media, "ended", onPlaybackChange, { signal });
		signal.addEventListener("abort", clearIdle, { once: true });
		scheduleIdle();
	}
});

//#endregion
export { controlsFeature };
