// @videojs/core/dist/default/dom/store/features/text-track.js
import { definePlayerFeature } from "../../feature.js";
import { findTrackElement, listen } from "@videojs/utils/dom";

//#region src/dom/store/features/text-track.ts
const textTrackFeature = definePlayerFeature({
	name: "textTrack",
	state: () => ({
		chaptersCues: [],
		thumbnailCues: [],
		thumbnailTrackSrc: null
	}),
	attach({ target, signal, set }) {
		const { media } = target;
		let trackCleanup = null;
		function sync() {
			trackCleanup?.abort();
			trackCleanup = new AbortController();
			let chaptersTrack = null;
			let thumbnailTrack = null;
			for (let i = 0; i < media.textTracks.length; i++) {
				const track = media.textTracks[i];
				if (!chaptersTrack && track.kind === "chapters") chaptersTrack = track;
				if (!thumbnailTrack && track.kind === "metadata" && track.label === "thumbnails") thumbnailTrack = track;
			}
			const chaptersCues = chaptersTrack?.cues ? Array.from(chaptersTrack.cues) : [];
			const thumbnailCues = thumbnailTrack?.cues ? Array.from(thumbnailTrack.cues) : [];
			let thumbnailTrackSrc = null;
			if (thumbnailTrack) thumbnailTrackSrc = findTrackElement(media, thumbnailTrack)?.src ?? null;
			for (const trackEl of media.querySelectorAll("track")) if (!trackEl.track?.cues?.length) listen(trackEl, "load", sync, { signal: trackCleanup.signal });
			set({
				chaptersCues,
				thumbnailCues,
				thumbnailTrackSrc
			});
		}
		sync();
		listen(media.textTracks, "addtrack", sync, { signal });
		listen(media.textTracks, "removetrack", sync, { signal });
		listen(media.textTracks, "change", sync, { signal });
		listen(media, "loadstart", sync, { signal });
		signal.addEventListener("abort", () => trackCleanup?.abort(), { once: true });
	}
});

//#endregion
export { textTrackFeature };
