// @videojs/core/dist/default/dom/store/features/time.js
import { definePlayerFeature } from "../../feature.js";
import { hasMetadata } from "../../media/predicate.js";
import { signalKeys } from "../signal-keys.js";
import { listen, onEvent } from "@videojs/utils/dom";
import { noop } from "@videojs/utils/function";

//#region src/dom/store/features/time.ts
const timeFeature = definePlayerFeature({
	name: "time",
	state: ({ target, signals }) => ({
		currentTime: 0,
		duration: 0,
		seeking: false,
		async seek(time) {
			const { media } = target(), signal = signals.supersede(signalKeys.seek);
			if (!hasMetadata(media)) {
				if (!await onEvent(media, "loadedmetadata", { signal }).catch(() => false)) return media.currentTime;
			}
			media.currentTime = Math.max(0, Math.min(time, media.duration || Infinity));
			await onEvent(media, "seeked", { signal }).catch(noop);
			return media.currentTime;
		}
	}),
	attach({ target, signal, set }) {
		const { media } = target;
		const sync = () => set({
			currentTime: media.currentTime,
			duration: Number.isFinite(media.duration) ? media.duration : 0,
			seeking: media.seeking
		});
		sync();
		listen(media, "timeupdate", sync, { signal });
		listen(media, "durationchange", sync, { signal });
		listen(media, "seeking", sync, { signal });
		listen(media, "seeked", sync, { signal });
		listen(media, "loadedmetadata", sync, { signal });
		listen(media, "emptied", sync, { signal });
	}
});

//#endregion
export { timeFeature };
