// @videojs/core/dist/default/dom/store/features/volume.js
import { definePlayerFeature } from "../../feature.js";
import { listen } from "@videojs/utils/dom";

//#region src/dom/store/features/volume.ts
const volumeFeature = definePlayerFeature({
	name: "volume",
	state: ({ target }) => ({
		volume: 1,
		muted: false,
		volumeAvailability: "unavailable",
		changeVolume(volume) {
			const { media } = target();
			media.volume = Math.max(0, Math.min(1, volume));
			return media.volume;
		},
		toggleMute() {
			const { media } = target();
			media.muted = !media.muted;
			return media.muted;
		}
	}),
	attach({ target, signal, set }) {
		const { media } = target;
		set({ volumeAvailability: canSetVolume() });
		const sync = () => set({
			volume: media.volume,
			muted: media.muted
		});
		sync();
		listen(media, "volumechange", sync, { signal });
	}
});
/** Check if volume can be programmatically set (fails on iOS Safari). */
function canSetVolume() {
	const video = document.createElement("video");
	try {
		video.volume = .5;
		return video.volume === .5 ? "available" : "unsupported";
	} catch {
		return "unsupported";
	}
}

//#endregion
export { volumeFeature };
