// @videojs/core/dist/dev/core/ui/fullscreen-button/fullscreen-button-core.js
import { defaults } from "@videojs/utils/object";
import { isFunction } from "@videojs/utils/predicate";

//#region src/core/ui/fullscreen-button/fullscreen-button-core.ts
var FullscreenButtonCore = class FullscreenButtonCore {
	static defaultProps = {
		label: "",
		disabled: false
	};
	#props = { ...FullscreenButtonCore.defaultProps };
	#media = null;
	constructor(props) {
		if (props) this.setProps(props);
	}
	setProps(props) {
		this.#props = defaults(props, FullscreenButtonCore.defaultProps);
	}
	getLabel(state) {
		const { label } = this.#props;
		if (isFunction(label)) {
			const customLabel = label(state);
			if (customLabel) return customLabel;
		} else if (label) return label;
		return state.fullscreen ? "Exit fullscreen" : "Enter fullscreen";
	}
	getAttrs(state) {
		return {
			"aria-label": this.getLabel(state),
			"aria-disabled": this.#props.disabled ? "true" : void 0
		};
	}
	setMedia(media) {
		this.#media = media;
	}
	getState() {
		const media = this.#media;
		return {
			fullscreen: media.fullscreen,
			availability: media.fullscreenAvailability
		};
	}
	async toggle(media) {
		if (this.#props.disabled) return;
		if (media.fullscreenAvailability !== "available") return;
		try {
			if (media.fullscreen) await media.exitFullscreen();
			else await media.requestFullscreen();
		} catch {}
	}
};

//#endregion
export { FullscreenButtonCore };
