// @videojs/core/dist/default/dom/presentation/fullscreen.js
import { isFunction } from "@videojs/utils/predicate";

//#region src/dom/presentation/fullscreen.ts
/** Check if the Fullscreen API is supported on this platform. */
function isFullscreenSupported() {
	const doc = document;
	if (doc.fullscreenEnabled || doc.webkitFullscreenEnabled) return true;
	return document.createElement("video").webkitSupportsFullscreen === true;
}
/** Get the current fullscreen element from the document. */
function getFullscreenElement() {
	const doc = document;
	return doc.fullscreenElement ?? doc.webkitFullscreenElement ?? null;
}
/**
* Check if a specific element (or its media) is currently in fullscreen.
*
* Uses `:fullscreen` pseudo-class which works across Shadow DOM boundaries.
*/
function isElementFullscreen(container, media) {
	const video = media;
	if (video.webkitDisplayingFullscreen && video.webkitPresentationMode === "fullscreen") return true;
	const target = container ?? media;
	if (getFullscreenElement() === target) return true;
	try {
		return target.matches(":fullscreen");
	} catch {
		return false;
	}
}
/**
* Enter fullscreen mode.
*
* Tries container first (to show custom UI), falls back to media element
* for platforms that only support video fullscreen (iOS Safari).
*/
async function enterFullscreen(container, media) {
	const video = media;
	if (container) {
		const el = container;
		if (isFunction(el.requestFullscreen)) return el.requestFullscreen();
		if (isFunction(el.webkitRequestFullscreen)) return el.webkitRequestFullscreen();
		if (isFunction(el.webkitRequestFullScreen)) return el.webkitRequestFullScreen();
	}
	if (isFunction(video.webkitEnterFullscreen)) {
		video.webkitEnterFullscreen();
		return;
	}
	if (isFunction(media.requestFullscreen)) return media.requestFullscreen();
	throw new DOMException("Fullscreen not supported", "NotSupportedError");
}
/** Exit fullscreen mode. */
async function exitFullscreen() {
	const doc = document;
	const video = getFullscreenElement();
	if (isFunction(doc.exitFullscreen)) return doc.exitFullscreen();
	if (isFunction(doc.webkitExitFullscreen)) return doc.webkitExitFullscreen();
	if (isFunction(doc.webkitCancelFullScreen)) return doc.webkitCancelFullScreen();
	if (video && isFunction(video.webkitExitFullscreen)) {
		video.webkitExitFullscreen();
		return;
	}
}

//#endregion
export { enterFullscreen, exitFullscreen, isElementFullscreen, isFullscreenSupported };
