// @videojs/react/dist/dev/presets/video/minimal-skin.tailwind.js
import { Container } from "../../player/context.js";
import { BufferingIndicator } from "../../ui/buffering-indicator/buffering-indicator.js";
import { ControlsRoot } from "../../ui/controls/controls-root.js";
import { FullscreenButton } from "../../ui/fullscreen-button/fullscreen-button.js";
import { MuteButton } from "../../ui/mute-button/mute-button.js";
import { PiPButton } from "../../ui/pip-button/pip-button.js";
import { PlayButton } from "../../ui/play-button/play-button.js";
import { PlaybackRateButton } from "../../ui/playback-rate-button/playback-rate-button.js";
import { PopoverPopup } from "../../ui/popover/popover-popup.js";
import { PopoverRoot } from "../../ui/popover/popover-root.js";
import { PopoverTrigger } from "../../ui/popover/popover-trigger.js";
import { SeekButton } from "../../ui/seek-button/seek-button.js";
import { SliderBuffer } from "../../ui/slider/slider-buffer.js";
import { SliderFill as SliderFill$1 } from "../../ui/slider/slider-fill.js";
import { SliderThumb as SliderThumb$1 } from "../../ui/slider/slider-thumb.js";
import { SliderTrack as SliderTrack$1 } from "../../ui/slider/slider-track.js";
import { Group } from "../../ui/time/time-group.js";
import { Separator } from "../../ui/time/time-separator.js";
import { Value } from "../../ui/time/time-value.js";
import { TimeSliderRoot } from "../../ui/time-slider/time-slider-root.js";
import { VolumeSliderRoot } from "../../ui/volume-slider/volume-slider-root.js";
import CaptionsOffIcon from "../../icons/dist/react/minimal/captions-off.js";
import CaptionsOnIcon from "../../icons/dist/react/minimal/captions-on.js";
import FullscreenEnterIcon from "../../icons/dist/react/minimal/fullscreen-enter.js";
import FullscreenExitIcon from "../../icons/dist/react/minimal/fullscreen-exit.js";
import PauseIcon from "../../icons/dist/react/minimal/pause.js";
import PipIcon from "../../icons/dist/react/minimal/pip.js";
import PlayIcon from "../../icons/dist/react/minimal/play.js";
import RestartIcon from "../../icons/dist/react/minimal/restart.js";
import SeekIcon from "../../icons/dist/react/minimal/seek.js";
import SpinnerIcon from "../../icons/dist/react/minimal/spinner.js";
import VolumeHighIcon from "../../icons/dist/react/minimal/volume-high.js";
import VolumeLowIcon from "../../icons/dist/react/minimal/volume-low.js";
import VolumeOffIcon from "../../icons/dist/react/minimal/volume-off.js";
import { CaptionsButton } from "../../ui/captions-button/captions-button.js";
import { ErrorDialog } from "../../ui/error-dialog/error-dialog.js";
import { playbackRate } from "../../skins/dist/default/video/default.tailwind.js";
import { bufferingIndicator, button, buttonGroup, controls, error, icon, iconContainer, iconFlipped, iconState, overlay, popup, root, seek, slider, time } from "../../skins/dist/default/video/minimal.tailwind.js";
import { forwardRef } from "react";
import { jsx, jsxs } from "react/jsx-runtime";
import { cn } from "@videojs/utils/style";

//#region src/presets/video/minimal-skin.tailwind.tsx
const SEEK_TIME = 10;
const Button = forwardRef(function Button({ className, variant, ...props }, ref) {
	return /* @__PURE__ */ jsx("button", {
		ref,
		type: "button",
		className: cn(button.base, variant === "icon" ? button.icon : button.default, className),
		...props
	});
});
const SliderRoot = forwardRef(function SliderRoot({ className, ...props }, ref) {
	return /* @__PURE__ */ jsx("div", {
		ref,
		className: cn(slider.root, className),
		...props
	});
});
const SliderTrack = forwardRef(function SliderTrack({ className, ...props }, ref) {
	return /* @__PURE__ */ jsx("div", {
		ref,
		className: cn(slider.track, className),
		...props
	});
});
const SliderFill = forwardRef(function SliderFill({ type = "fill", className, ...props }, ref) {
	return /* @__PURE__ */ jsx("div", {
		ref,
		className: cn(slider.fill.base, type === "fill" ? slider.fill.fill : slider.fill.buffer, className),
		...props
	});
});
const SliderThumb = forwardRef(function SliderThumb({ persistent, className, ...props }, ref) {
	return /* @__PURE__ */ jsx("div", {
		ref,
		className: cn(slider.thumb.base, persistent ? void 0 : slider.thumb.interactive, className),
		...props
	});
});
function MinimalVideoSkinTailwind(props) {
	const { children, className, ...rest } = props;
	return /* @__PURE__ */ jsxs(Container, {
		className: cn(root, className),
		...rest,
		children: [
			/* @__PURE__ */ jsx(BufferingIndicator, { render: (props) => /* @__PURE__ */ jsx("div", {
				...props,
				className: bufferingIndicator,
				children: /* @__PURE__ */ jsx(SpinnerIcon, { className: icon })
			}) }),
			/* @__PURE__ */ jsx(ErrorDialog, {
				"aria-labelledby": "media-error-title",
				"aria-describedby": "media-error-description",
				render: (props, { onDismiss }) => /* @__PURE__ */ jsx("div", {
					...props,
					className: error.root,
					children: /* @__PURE__ */ jsxs("div", {
						className: error.dialog,
						children: [/* @__PURE__ */ jsxs("div", {
							className: error.content,
							children: [/* @__PURE__ */ jsx("p", {
								id: "media-error-title",
								className: error.title,
								children: "Something went wrong."
							}), /* @__PURE__ */ jsx("p", {
								id: "media-error-description",
								children: "An error occurred while trying to play the video. Please try again."
							})]
						}), /* @__PURE__ */ jsx("div", {
							className: error.actions,
							children: /* @__PURE__ */ jsx(Button, {
								onClick: onDismiss,
								children: "OK"
							})
						})]
					})
				})
			}),
			/* @__PURE__ */ jsxs(ControlsRoot, {
				"data-controls": "",
				className: controls,
				children: [
					/* @__PURE__ */ jsxs("span", {
						className: buttonGroup,
						children: [
							/* @__PURE__ */ jsx(PlayButton, { render: (props) => /* @__PURE__ */ jsxs(Button, {
								variant: "icon",
								...props,
								className: iconState.play.button,
								children: [
									/* @__PURE__ */ jsx(RestartIcon, { className: cn(icon, iconState.play.restart) }),
									/* @__PURE__ */ jsx(PlayIcon, { className: cn(icon, iconState.play.play) }),
									/* @__PURE__ */ jsx(PauseIcon, { className: cn(icon, iconState.play.pause) })
								]
							}) }),
							/* @__PURE__ */ jsx(SeekButton, {
								seconds: -SEEK_TIME,
								render: (props) => /* @__PURE__ */ jsx(Button, {
									variant: "icon",
									...props,
									className: seek.button,
									children: /* @__PURE__ */ jsxs("span", {
										className: iconContainer,
										children: [/* @__PURE__ */ jsx(SeekIcon, { className: cn(icon, iconFlipped) }), /* @__PURE__ */ jsx("span", {
											className: cn(seek.label, seek.labelBackward),
											children: SEEK_TIME
										})]
									})
								})
							}),
							/* @__PURE__ */ jsx(SeekButton, {
								seconds: SEEK_TIME,
								render: (props) => /* @__PURE__ */ jsx(Button, {
									variant: "icon",
									...props,
									className: seek.button,
									children: /* @__PURE__ */ jsxs("span", {
										className: iconContainer,
										children: [/* @__PURE__ */ jsx(SeekIcon, { className: icon }), /* @__PURE__ */ jsx("span", {
											className: cn(seek.label, seek.labelForward),
											children: SEEK_TIME
										})]
									})
								})
							})
						]
					}),
					/* @__PURE__ */ jsxs("span", {
						className: time.controls,
						children: [/* @__PURE__ */ jsxs(Group, {
							className: time.group,
							children: [
								/* @__PURE__ */ jsx(Value, {
									type: "current",
									className: time.current
								}),
								/* @__PURE__ */ jsx(Separator, { className: time.separator }),
								/* @__PURE__ */ jsx(Value, {
									type: "duration",
									className: time.duration
								})
							]
						}), /* @__PURE__ */ jsxs(TimeSliderRoot, {
							render: (props) => /* @__PURE__ */ jsx(SliderRoot, { ...props }),
							children: [/* @__PURE__ */ jsxs(SliderTrack$1, {
								render: (props) => /* @__PURE__ */ jsx(SliderTrack, { ...props }),
								children: [/* @__PURE__ */ jsx(SliderFill$1, { render: (props) => /* @__PURE__ */ jsx(SliderFill, { ...props }) }), /* @__PURE__ */ jsx(SliderBuffer, { render: (props) => /* @__PURE__ */ jsx(SliderFill, {
									type: "buffer",
									...props
								}) })]
							}), /* @__PURE__ */ jsx(SliderThumb$1, { render: (props) => /* @__PURE__ */ jsx(SliderThumb, { ...props }) })]
						})]
					}),
					/* @__PURE__ */ jsxs("span", {
						className: buttonGroup,
						children: [
							/* @__PURE__ */ jsx(PlaybackRateButton, { render: (props) => /* @__PURE__ */ jsx(Button, {
								variant: "icon",
								...props,
								className: playbackRate.button
							}) }),
							/* @__PURE__ */ jsxs(PopoverRoot, {
								openOnHover: true,
								delay: 200,
								closeDelay: 100,
								side: "top",
								children: [/* @__PURE__ */ jsx(PopoverTrigger, { render: /* @__PURE__ */ jsx(MuteButton, { render: (props) => /* @__PURE__ */ jsxs(Button, {
									variant: "icon",
									...props,
									className: iconState.mute.button,
									children: [
										/* @__PURE__ */ jsx(VolumeOffIcon, { className: cn(icon, iconState.mute.volumeOff) }),
										/* @__PURE__ */ jsx(VolumeLowIcon, { className: cn(icon, iconState.mute.volumeLow) }),
										/* @__PURE__ */ jsx(VolumeHighIcon, { className: cn(icon, iconState.mute.volumeHigh) })
									]
								}) }) }), /* @__PURE__ */ jsx(PopoverPopup, {
									className: cn(popup.base, popup.volume),
									children: /* @__PURE__ */ jsxs(VolumeSliderRoot, {
										orientation: "vertical",
										thumbAlignment: "edge",
										render: (props) => /* @__PURE__ */ jsx(SliderRoot, { ...props }),
										children: [/* @__PURE__ */ jsx(SliderTrack$1, {
											render: (props) => /* @__PURE__ */ jsx(SliderTrack, { ...props }),
											children: /* @__PURE__ */ jsx(SliderFill$1, { render: (props) => /* @__PURE__ */ jsx(SliderFill, { ...props }) })
										}), /* @__PURE__ */ jsx(SliderThumb$1, { render: (props) => /* @__PURE__ */ jsx(SliderThumb, {
											persistent: true,
											...props
										}) })]
									})
								})]
							}),
							/* @__PURE__ */ jsx(CaptionsButton, { render: (props) => /* @__PURE__ */ jsxs(Button, {
								variant: "icon",
								...props,
								className: iconState.captions.button,
								children: [/* @__PURE__ */ jsx(CaptionsOffIcon, { className: cn(icon, iconState.captions.off) }), /* @__PURE__ */ jsx(CaptionsOnIcon, { className: cn(icon, iconState.captions.on) })]
							}) }),
							/* @__PURE__ */ jsx(PiPButton, { render: (props) => /* @__PURE__ */ jsx(Button, {
								variant: "icon",
								...props,
								children: /* @__PURE__ */ jsx(PipIcon, { className: icon })
							}) }),
							/* @__PURE__ */ jsx(FullscreenButton, { render: (props) => /* @__PURE__ */ jsxs(Button, {
								variant: "icon",
								...props,
								className: iconState.fullscreen.button,
								children: [/* @__PURE__ */ jsx(FullscreenEnterIcon, { className: cn(icon, iconState.fullscreen.enter) }), /* @__PURE__ */ jsx(FullscreenExitIcon, { className: cn(icon, iconState.fullscreen.exit) })]
							}) })
						]
					})
				]
			}),
			/* @__PURE__ */ jsx("div", { className: overlay }),
			children
		]
	});
}

//#endregion
export { MinimalVideoSkinTailwind };
