// @videojs/react/dist/dev/ui/popover/popover-trigger.js
'use client';

import { renderElement } from "../../utils/use-render.js";
import { usePopoverContext } from "./context.js";
import { getAnchorNameStyle } from "@videojs/core/dom";
import { forwardRef, useCallback, useMemo } from "react";

//#region src/ui/popover/popover-trigger.tsx
/** Button that toggles the popover visibility. Renders a `<button>` element. */
const PopoverTrigger = forwardRef(function PopoverTrigger({ render, className, style, ...elementProps }, forwardedRef) {
	const { core, popover, state, stateAttrMap, anchorName, popupId } = usePopoverContext();
	const triggerRef = useCallback((el) => {
		popover.setTriggerElement(el);
	}, [popover]);
	const anchorStyle = useMemo(() => getAnchorNameStyle(anchorName), [anchorName]);
	const { onFocusIn, onFocusOut, ...restTriggerProps } = popover.triggerProps;
	return renderElement("button", {
		render,
		className,
		style
	}, {
		state,
		stateAttrMap,
		ref: [forwardedRef, triggerRef],
		props: [
			{
				type: "button",
				style: anchorStyle,
				...core.getTriggerAttrs(state, popupId)
			},
			{
				...restTriggerProps,
				onFocus: onFocusIn,
				onBlur: onFocusOut
			},
			elementProps
		]
	});
});

//#endregion
export { PopoverTrigger };
