// @videojs/core/dist/default/dom/store/selectors.js
import { bufferFeature } from "./features/buffer.js";
import { controlsFeature } from "./features/controls.js";
import { fullscreenFeature } from "./features/fullscreen.js";
import { pipFeature } from "./features/pip.js";
import { playbackFeature } from "./features/playback.js";
import { playbackRateFeature } from "./features/playback-rate.js";
import { sourceFeature } from "./features/source.js";
import { textTrackFeature } from "./features/text-track.js";
import { timeFeature } from "./features/time.js";
import { volumeFeature } from "./features/volume.js";
import { createSelector } from "@videojs/store";

//#region src/dom/store/selectors.ts
/** Select the buffer state (buffered ranges, percent buffered). */
const selectBuffer = createSelector(bufferFeature);
/** Select the controls state (controls visible, user-active). */
const selectControls = createSelector(controlsFeature);
/** Select the fullscreen state (fullscreen active, availability). */
const selectFullscreen = createSelector(fullscreenFeature);
/** Select the PiP state (picture-in-picture active, availability). */
const selectPiP = createSelector(pipFeature);
/** Select the playback state (paused, ended, play, pause, toggle). */
const selectPlayback = createSelector(playbackFeature);
/** Select the playback rate state (playbackRate, playbackRates, setPlaybackRate). */
const selectPlaybackRate = createSelector(playbackRateFeature);
/** Select the source state (src, type). */
const selectSource = createSelector(sourceFeature);
/** Select the text track state (chapters cues, thumbnail cues). */
const selectTextTrack = createSelector(textTrackFeature);
/** Select the time state (currentTime, duration, seek). */
const selectTime = createSelector(timeFeature);
/** Select the volume state (volume, muted, setVolume, setMuted). */
const selectVolume = createSelector(volumeFeature);

//#endregion
export { selectBuffer, selectControls, selectFullscreen, selectPiP, selectPlayback, selectPlaybackRate, selectSource, selectTextTrack, selectTime, selectVolume };
