// @videojs/core/dist/default/core/media/proxy.js
//#region src/core/media/proxy.ts
const API_METHOD = 0;
const API_GET = 1;
const API_GET_SET = 2;
/**
* This mixin creates an API from the passed classes and proxies the methods and properties to the attached target.
*
* Many methods and properties will need no translation and are proxied directly to the attached target.
* For example, the `play` and `pause` methods are proxied directly to the attached target.
*
* Child classes can override the proxied methods and properties to provide custom behavior.
* For example, the `src` property for HLS media is proxied to the HLS engine, not the target itself.
*
* The `get`, `set`, and `call` methods can be overridden to provide catch-all custom behavior.
*/
const MediaProxyMixin = (PrimaryClass, ...AdditionalClasses) => {
	class MediaApiProxy {
		static extends(...MediaApiTargetClasses) {
			const props = getClassProps(...MediaApiTargetClasses);
			for (const [prop, type] of props.entries()) {
				if (prop in MediaApiProxy.prototype) continue;
				const config = {};
				if (type === API_METHOD) config.value = function(...args) {
					return this.call(prop, ...args);
				};
				else if (type === API_GET || type === API_GET_SET) {
					config.get = function() {
						return this.get(prop);
					};
					if (type === API_GET_SET) config.set = function(val) {
						this.set(prop, val);
					};
				}
				Object.defineProperty(MediaApiProxy.prototype, prop, config);
			}
		}
		#target = null;
		get target() {
			return this.#target;
		}
		get(prop) {
			return this.target?.[prop];
		}
		set(prop, val) {
			if (this.target) this.target[prop] = val;
		}
		call(prop, ...args) {
			return (this.target?.[prop])?.apply(this.target, args);
		}
		attach(target) {
			if (!target || this.#target === target) return;
			this.#target = target;
		}
		detach() {
			if (!this.#target) return;
			this.#target = null;
		}
	}
	MediaApiProxy.extends(PrimaryClass, ...AdditionalClasses);
	return MediaApiProxy;
};
/**
* Helper function to get the methods, getters, and setters from a class prototype.
*/
function getClassProps(...Classes) {
	const props = /* @__PURE__ */ new Map();
	for (const Class of Classes) {
		const names = Object.getOwnPropertyNames(Class.prototype);
		for (const name of names) {
			const descriptor = Object.getOwnPropertyDescriptor(Class.prototype, name);
			if (typeof descriptor?.value === "function") props.set(name, API_METHOD);
			else if (typeof descriptor?.get === "function") if (typeof descriptor?.set === "function") props.set(name, API_GET_SET);
			else props.set(name, API_GET);
		}
	}
	return props;
}

//#endregion
export { MediaProxyMixin, getClassProps };
