// @videojs/react/dist/default/ui/create-media-button.js
'use client';

import { usePlayer } from "../player/context.js";
import { renderElement } from "../utils/use-render.js";
import { useButton } from "./hooks/use-button.js";
import "@videojs/core/dom";
import { forwardRef, useState } from "react";

//#region src/ui/create-media-button.tsx
/** Creates a media button React component from a core class and config. */
function createMediaButton(config) {
	const { displayName, core: CoreClass, stateAttrMap, selector, action } = config;
	const corePropKeys = new Set(Object.keys(CoreClass.defaultProps));
	const Component = forwardRef(function MediaButton(componentProps, forwardedRef) {
		const { render, className, style, ...rest } = componentProps;
		const coreProps = {};
		const elementProps = {};
		for (const key of Object.keys(rest)) if (corePropKeys.has(key)) coreProps[key] = rest[key];
		else elementProps[key] = rest[key];
		const feature = usePlayer(selector);
		const [core] = useState(() => new CoreClass());
		core.setProps(coreProps);
		const { getButtonProps, buttonRef } = useButton({
			displayName,
			onActivate: () => action(core, feature),
			isDisabled: () => !!coreProps.disabled || !feature
		});
		if (!feature) return null;
		core.setMedia(feature);
		const state = core.getState();
		return renderElement("button", {
			render,
			className,
			style
		}, {
			state,
			stateAttrMap,
			ref: [forwardedRef, buttonRef],
			props: [
				core.getAttrs(state),
				elementProps,
				getButtonProps()
			]
		});
	});
	Component.displayName = displayName;
	return Component;
}

//#endregion
export { createMediaButton };
