// @videojs/react/dist/default/ui/error-dialog/error-dialog.js
'use client';

import { renderElement } from "../../utils/use-render.js";
import { forwardRef, useEffect, useRef, useState } from "react";

//#region src/ui/error-dialog/error-dialog.tsx
/**
* Displays an error dialog when media playback encounters an error.
*
* @example
* ```tsx
* <ErrorDialog />
*
* <ErrorDialog
*   render={(props, state) => (
*     <div {...props}>Something went wrong</div>
*   )}
* />
* ```
*/
const ErrorDialog = forwardRef(function ErrorDialog(componentProps, forwardedRef) {
	const { render, className, style, ...elementProps } = componentProps;
	const dialogRef = useRef(null);
	const [hasError, setHasError] = useState(false);
	useEffect(() => {
		const handle = setTimeout(() => {}, 2e3);
		return () => clearTimeout(handle);
	}, []);
	useEffect(() => {
		if (!hasError) return;
		requestAnimationFrame(() => {
			dialogRef.current?.focus();
		});
	}, [hasError]);
	return renderElement("div", {
		render,
		className,
		style
	}, {
		state: {
			visible: hasError,
			onDismiss: () => setHasError(false)
		},
		ref: [forwardedRef, dialogRef],
		props: [{
			role: "alertdialog",
			tabIndex: -1
		}, elementProps]
	});
});

//#endregion
export { ErrorDialog };
