// @videojs/core/dist/dev/dom/ui/popover/popover.js
import { listen } from "@videojs/utils/dom";

//#region src/dom/ui/popover/popover.ts
function createPopover(options) {
	const { transition, onOpenChange, closeOnEscape, closeOnOutsideClick } = options;
	const state = transition.state;
	let triggerEl = null;
	let popupEl = null;
	let hoverTimeout = null;
	const abort = new AbortController();
	let docAc = null;
	function clearHoverTimeout() {
		if (hoverTimeout !== null) {
			clearTimeout(hoverTimeout);
			hoverTimeout = null;
		}
	}
	function canHover() {
		return globalThis.matchMedia?.("(hover: hover)")?.matches ?? false;
	}
	/**
	* The transition handler manages animation lifecycle via `createState`:
	*
	* **Open:** `transition.open()` patches `{ active: true, status: 'starting' }`.
	* After one RAF it patches `{ status: 'idle' }` and the promise resolves.
	* Frameworks render `data-starting-style` / `data-ending-style` via
	* `getPopupAttrs(state)` — no imperative DOM mutation needed.
	*
	* **Close:** `transition.close(el)` patches `{ status: 'ending' }` (keeping
	* `active: true` so the element stays mounted). After a double-RAF it waits
	* for `getAnimations()` to settle, then patches `{ active: false, status: 'idle' }`.
	*
	* `onOpenChange` fires immediately (before animations).
	* `onOpenChangeComplete` fires after animations finish.
	*/
	function applyOpen(reason, event) {
		if (abort.signal.aborted) return;
		const { active, status } = state.current;
		if (active && status !== "ending") return;
		if (status === "ending") transition.cancel();
		transition.open().then(() => {
			if (abort.signal.aborted || !state.current.active) return;
			options.onOpenChangeComplete?.(true);
		});
		tryShowPopover(popupEl);
		onOpenChange(true, event ? {
			reason,
			event
		} : { reason });
	}
	function applyClose(reason, event) {
		const { active, status } = state.current;
		if (abort.signal.aborted || !active || status === "ending") return;
		transition.close(popupEl).then(() => {
			if (abort.signal.aborted) return;
			tryHidePopover(popupEl);
			options.onOpenChangeComplete?.(false);
		});
		onOpenChange(false, event ? {
			reason,
			event
		} : { reason });
	}
	function open(reason = "click") {
		applyOpen(reason);
	}
	function close(reason = "click") {
		applyClose(reason);
	}
	function setupDocumentListeners() {
		cleanupDocumentListeners();
		if (typeof document === "undefined") return;
		docAc = new AbortController();
		const signal = docAc.signal;
		listen(document, "keydown", handleDocumentKeydown, { signal });
		listen(document, "pointerdown", handleDocumentPointerdown, {
			capture: true,
			signal
		});
	}
	function cleanupDocumentListeners() {
		docAc?.abort();
		docAc = null;
	}
	function handleDocumentKeydown(event) {
		if (event.key === "Escape" && closeOnEscape() && state.current.active) {
			event.preventDefault();
			applyClose("escape", event);
		}
	}
	function handleDocumentPointerdown(event) {
		if (!closeOnOutsideClick() || !state.current.active) return;
		const target = event.target;
		if (!target) return;
		if (triggerEl?.contains(target) || popupEl?.contains(target)) return;
		applyClose("outside-click", event);
	}
	const unsubscribe = state.subscribe(() => {
		if (state.current.active) setupDocumentListeners();
		else cleanupDocumentListeners();
	});
	abort.signal.addEventListener("abort", () => {
		unsubscribe();
		clearHoverTimeout();
		transition.destroy();
		cleanupDocumentListeners();
		triggerEl = null;
		popupEl = null;
	});
	const triggerProps = {
		onClick(event) {
			if (state.current.active && state.current.status !== "ending") applyClose("click", event);
			else applyOpen("click", event);
		},
		onPointerEnter(_event) {
			if (!options.openOnHover?.()) return;
			if (!canHover()) return;
			clearHoverTimeout();
			if (state.current.active) return;
			const delay = options.delay?.() ?? 300;
			hoverTimeout = setTimeout(() => applyOpen("hover"), delay);
		},
		onPointerLeave(_event) {
			if (!options.openOnHover?.()) return;
			if (!canHover()) return;
			clearHoverTimeout();
			if (!state.current.active) return;
			const closeDelay = options.closeDelay?.() ?? 0;
			hoverTimeout = setTimeout(() => applyClose("hover"), closeDelay);
		},
		onFocusIn(_event) {
			if (options.openOnHover?.()) applyOpen("focus");
		},
		onFocusOut(event) {
			const relatedTarget = event.relatedTarget;
			if (relatedTarget && (triggerEl?.contains(relatedTarget) || popupEl?.contains(relatedTarget))) return;
			if (options.openOnHover?.()) applyClose("blur");
		}
	};
	const popupProps = {
		onPointerEnter(_event) {
			if (!options.openOnHover?.()) return;
			clearHoverTimeout();
		},
		onPointerLeave(_event) {
			if (!options.openOnHover?.()) return;
			clearHoverTimeout();
			if (!state.current.active) return;
			const closeDelay = options.closeDelay?.() ?? 0;
			hoverTimeout = setTimeout(() => applyClose("hover"), closeDelay);
		},
		onFocusOut(event) {
			const relatedTarget = event.relatedTarget;
			if (relatedTarget && (triggerEl?.contains(relatedTarget) || popupEl?.contains(relatedTarget))) return;
			applyClose("blur");
		}
	};
	function setTriggerElement(el) {
		triggerEl = el;
	}
	function setPopupElement(el) {
		if (!el && popupEl && state.current.active) tryHidePopover(popupEl);
		popupEl = el;
		if (el) {
			if (state.current.active) tryShowPopover(el);
		}
	}
	function destroy() {
		if (abort.signal.aborted) return;
		abort.abort();
	}
	return {
		input: state,
		triggerProps,
		popupProps,
		get triggerElement() {
			return triggerEl;
		},
		setTriggerElement,
		setPopupElement,
		open,
		close,
		destroy
	};
}
function tryShowPopover(el) {
	try {
		el?.showPopover?.();
	} catch {}
}
function tryHidePopover(el) {
	try {
		el?.hidePopover?.();
	} catch {}
}

//#endregion
export { createPopover };
