// @videojs/core/dist/dev/dom/store/features/fullscreen.js
import { definePlayerFeature } from "../../feature.js";
import { enterFullscreen, exitFullscreen, isElementFullscreen, isFullscreenSupported } from "../../presentation/fullscreen.js";
import { exitPiP, isPiPActive } from "../../presentation/pip.js";
import { listen } from "@videojs/utils/dom";

//#region src/dom/store/features/fullscreen.ts
const fullscreenFeature = definePlayerFeature({
	name: "fullscreen",
	state: ({ target }) => ({
		fullscreen: false,
		fullscreenAvailability: "unavailable",
		async requestFullscreen() {
			const { media, container } = target();
			if (isPiPActive(media)) await exitPiP(media);
			return enterFullscreen(container, media);
		},
		async exitFullscreen() {
			return exitFullscreen();
		}
	}),
	attach({ target, signal, set }) {
		const { media, container } = target;
		set({ fullscreenAvailability: isFullscreenSupported() ? "available" : "unsupported" });
		const sync = () => set({ fullscreen: isElementFullscreen(container, media) });
		sync();
		listen(document, "fullscreenchange", sync, { signal });
		listen(document, "webkitfullscreenchange", sync, { signal });
		if ("webkitPresentationMode" in media) listen(media, "webkitpresentationmodechanged", sync, { signal });
	}
});

//#endregion
export { fullscreenFeature };
