// @videojs/core/dist/dev/dom/store/features/pip.js
import { definePlayerFeature } from "../../feature.js";
import { exitFullscreen, isElementFullscreen } from "../../presentation/fullscreen.js";
import { enterPiP, exitPiP, isPiPActive, isPiPSupported } from "../../presentation/pip.js";
import { listen } from "@videojs/utils/dom";

//#region src/dom/store/features/pip.ts
const pipFeature = definePlayerFeature({
	name: "pip",
	state: ({ target }) => ({
		pip: false,
		pipAvailability: "unavailable",
		async requestPiP() {
			const { media, container } = target();
			if (isElementFullscreen(container, media)) await exitFullscreen();
			return enterPiP(media);
		},
		async exitPiP() {
			const { media } = target();
			return exitPiP(media);
		}
	}),
	attach({ target, signal, set }) {
		const { media } = target;
		set({ pipAvailability: isPiPSupported() ? "available" : "unsupported" });
		const sync = () => set({ pip: isPiPActive(media) });
		sync();
		listen(media, "enterpictureinpicture", sync, { signal });
		listen(media, "leavepictureinpicture", sync, { signal });
		if ("webkitPresentationMode" in media) listen(media, "webkitpresentationmodechanged", sync, { signal });
	}
});

//#endregion
export { pipFeature };
