// @videojs/react/dist/default/ui/time/time-value.js
'use client';

import { usePlayer } from "../../player/context.js";
import { renderElement } from "../../utils/use-render.js";
import { selectTime } from "@videojs/core/dom";
import { forwardRef, useState } from "react";
import { Fragment, jsx, jsxs } from "react/jsx-runtime";
import { TimeCore, TimeDataAttrs } from "@videojs/core";

//#region src/ui/time/time-value.tsx
/**
* Displays a formatted time value (current, duration, or remaining).
*
* @example
* ```tsx
* <Time.Value />
* <Time.Value type="duration" />
* <Time.Value type="remaining" negativeSign="−" />
* ```
*/
const Value = forwardRef(function Value(componentProps, forwardedRef) {
	const { render, className, style, type, negativeSign, label, ...elementProps } = componentProps;
	const time = usePlayer(selectTime);
	const [core] = useState(() => new TimeCore());
	core.setProps({
		type,
		negativeSign,
		label
	});
	if (!time) return null;
	core.setMedia(time);
	const state = core.getState();
	const content = state.negative ? /* @__PURE__ */ jsxs(Fragment, { children: [/* @__PURE__ */ jsx("span", {
		"aria-hidden": "true",
		children: negativeSign ?? TimeCore.defaultProps.negativeSign
	}), state.text] }) : state.text;
	return renderElement("time", {
		render,
		className,
		style
	}, {
		state,
		stateAttrMap: TimeDataAttrs,
		ref: [forwardedRef],
		props: [{
			dateTime: state.datetime,
			children: content,
			...core.getAttrs(state)
		}, elementProps]
	});
});

//#endregion
export { Value };
